rcgen/src/string_types.rs

```diff
@@ -0,0 +1,490 @@
+///
+/// For the full ASCII character set, use
+/// [`Ia5String`][`crate::Ia5String`].
+///
+/// # Examples
+///
+/// You can create a `PrintableString` from [a literal string][`&str`] with [`PrintableString::try_from`]:
+///
+/// ```
+/// use rcgen::PrintableString;
+/// let hello = PrintableString::try_from("hello").unwrap();
+/// ```
+///
+/// # Supported characters
+///
+/// The following ASCII characters/ranges are supported:
+///
+/// - `A..Z`
+/// - `a..z`
+/// - `0..9`
+/// - "` `" (i.e. space)
+/// - `\`
+/// - `(`
+/// - `)`
+/// - `+`
+/// - `,`
+/// - `-`
+/// - `.`
+/// - `/`
+/// - `:`
+/// - `=`
+/// - `?`
+#[derive(Debug, PartialEq, Eq, Hash, Clone)]
+pub struct PrintableString(String);
+
+impl PrintableString {
+	/// Extracts a string slice containing the entire `PrintableString`.
+	pub fn as_str(&self) -> &str {
+		&self.0
+	}
+}
+
+impl TryFrom<&str> for PrintableString {
+	type Error = Error;
+
+	/// Converts a `&str` to a [`PrintableString`].
+	///
+	/// Any character not in the [`PrintableString`] charset will be rejected.
+	/// See [`PrintableString`] documentation for more information.
+	///
+	/// The result is allocated on the heap.
+	fn try_from(input: &str) -> Result<Self, Error> {
+		input.to_string().try_into()
+	}
+}
+
+impl TryFrom<String> for PrintableString {
+	type Error = Error;
+
+	/// Converts a [`String`][`std::string::String`] into a [`PrintableString`]
+	///
+	/// Any character not in the [`PrintableString`] charset will be rejected.
+	/// See [`PrintableString`] documentation for more information.
+	///
+	/// This conversion does not allocate or copy memory.
+	fn try_from(value: String) -> Result<Self, Self::Error> {
+		for &c in value.as_bytes() {
+			match c {
+				b'A'..=b'Z'
+				| b'a'..=b'z'
+				| b'0'..=b'9'
+				| b' '
+				| b'\''
+				| b'('
+				| b')'
+				| b'+'
+				| b','
+				| b'-'
+				| b'.'
+				| b'/'
+				| b':'
+				| b'='
+				| b'?' => (),
+				_ => return Err(Error::InvalidPrintableString),
+			}
+		}
+		Ok(Self(value))
+	}
+}
+
+impl FromStr for PrintableString {
+	type Err = Error;
+
+	fn from_str(s: &str) -> Result<Self, Self::Err> {
+		s.try_into()
+	}
+}
+
+impl AsRef<str> for PrintableString {
+	fn as_ref(&self) -> &str {
+		&self.0
+	}
+}
+
+/// ASN.1 `IA5String` type.
+///
+/// # Examples
+///
+/// You can create a `Ia5String` from [a literal string][`&str`] with [`Ia5String::try_from`]:
+///
+/// ```
+/// use rcgen::Ia5String;
+/// let hello = Ia5String::try_from("hello").unwrap();
+/// ```
+///
+/// # Supported characters
+///
+/// Supports the [International Alphabet No. 5 (IA5)] character encoding, i.e.
+/// the lower 128 characters of the ASCII alphabet. (Note: IA5 is now
+/// technically known as the International Reference Alphabet or IRA as
+/// specified in the ITU-T's T.50 recommendation).
+///
+/// For UTF-8, use [`String`][`std::string::String`].
+///
+/// [International Alphabet No. 5 (IA5)]: https://en.wikipedia.org/wiki/T.50_%28standard%2
```

Comment:
Does this link work? I tried to copy it, it doesn't work for me.

Reply:
I fix the link ;)